### PR TITLE
Implementation of MobileNetv2 and MobileNetv3

### PR DESCRIPTION
**Pull request by theabhirath**
This is an implementation of the two most recent and widely-used MobileNet models, [MobileNetv2](https://arxiv.org/abs/1801.04381) and [MobileNetv3](https://arxiv.org/pdf/1905.02244).

### CODE REVIEW

**Comment by lorenzoh** [comment]
Very nice! Have you tested the models on GPU? I recall trying to implement these a way back and the grouped convolutions not working yet.

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:91]:
```suggestion
  return Chain(Chain(layers...,
                     conv_bn((1, 1), inplanes, outplanes, relu6, bias = false)...),
               Chain(AdaptiveMeanPool((1, 1)), flatten, Dense(outplanes, nclasses)))
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:70]: Even though MobileNetV2 is a restricted architecture, should we allow this config to be passed in?

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:87]: Similarly, should we let the 1280 be customizable?

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:132]: Makes more sense to make a PR to NNlib.jl with this.

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:201]: Prefer to use `Symbol`s instead of `String`s.

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:225]:
```suggestion
  output_channel = (mode == "large") ? 1280 : 1024
  output_channel = width_mult > 1.0 ? _make_divisible(output_channel * width_mult, 8) : output_channel
```

Comment (darsnack) [inline src/mobilenet.jl:225]: Similar comment here too about making the output channels customizable, then restricting it at a higher level in the API.

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:234]:
```suggestion
  classifier = (
    Dense(explanes, output_channel, h_swish),
    Dropout(0.2),
    Dense(output_channel, nclasses),
  )

  return Chain(Chain(layers...,
                     conv_bn((1, 1), inplanes, explanes, h_swish, bias = false)...),
               Chain(AdaptiveMeanPool((1, 1)), flatten, classifier...))
```

File: test/runtests.jl
Comment (darsnack) [inline test/runtests.jl:136]:
```suggestion
end

```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:218]: Is this so that you can access `explanes` outside the loop? Instead of using `global`, you should initialize it before the loop (e.g. `explanes = 0`). Or even better, leave it as is without `global`, then outside the loop use `Flux.outputsize` to get the size for use later in the code.

**Review by darsnack — CHANGES_REQUESTED** [review]
Looks pretty good on the first pass.

In general, for the lowest level function (e.g. `mobilenetv2`), we want to cast as wide a net as possible when it comes to configuration. If you find that it is possible to loop over a configuration dictionary, then it should probably be a user-specified parameter. The higher level functions like `MobileNetV2` exist to provide convenience constructors for passing in the correct configs.

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl:215]: Whoops that was a bit of hacky code I used because I copied the `configs` dict over and didn't want to spend time fixing the types 😅. I'll fix that up

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl:218]: Yeah I missed `Flux.outputsize`, that's an amazing function xD. Will make the change

**Comment by theabhirath** [comment]
> Looks pretty good on the first pass.
> 
> In general, for the lowest level function (e.g. `mobilenetv2`), we want to cast as wide a net as possible when it comes to configuration. If you find that it is possible to loop over a configuration dictionary, then it should probably be a user-specified parameter. The higher level functions like `MobileNetV2` exist to provide convenience constructors for passing in the correct configs.

Okay I'll fix the formatting errors (tried my best, some seem to have gotten past again). For the configurations, I initially thought of passing the `configs` in but I ultimately decided not to because:

1. It's a fairly large configuration and I'm not quite sure how we'd deal with all the edge cases for the possible values of the kernel size etc. and 
2. Almost every use case I found changed the width multiplier, not the configs because that's a more fundamental value to the network. 

Lemme know if you still think I should make that change though, I'll go ahead.

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl:132]: Have done so, will use that version once it gets merged

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl]: Have done. I'm a little new to the concept of metaprogramming though, so what exactly is the difference here? And why are `Symbols` preferred?

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl]: I've changed the API to allow for this and added a note in the documentation, hopefully that should prevent people from causing errors with the configurations

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl:133]: https://github.com/FluxML/NNlib.jl/pull/376 takes care of this. But given that it's in `master` and not in the current release, keeping this line of code for now.

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl]: Fixed this in the latest API

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl]: Fixed

**Comment by theabhirath** [comment]
> > Looks pretty good on the first pass.
> > In general, for the lowest level function (e.g. `mobilenetv2`), we want to cast as wide a net as possible when it comes to configuration. If you find that it is possible to loop over a configuration dictionary, then it should probably be a user-specified parameter. The higher level functions like `MobileNetV2` exist to provide convenience constructors for passing in the correct configs.
> 
> Okay I'll fix the formatting errors (tried my best, some seem to have gotten past again). For the configurations, I initially thought of passing the `configs` in but I ultimately decided not to because:
> 
> 1. It's a fairly large configuration and I'm not quite sure how we'd deal with all the edge cases for the possible values of the kernel size etc. and
> 2. Almost every use case I found changed the width multiplier, not the configs because that's a more fundamental value to the network.
> 
> Lemme know if you still think I should make that change though, I'll go ahead.

Decided to just document it well and hope that the users don't try and do something too dangerous with the models 😂. There's examples in the code though so hopefully it isn't very difficult to follow

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:52]:
```suggestion
- `configs`: A "vector of vectors" configuration for each layer that details:
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:51]:
```suggestion
- `width_mult`: Controls the number of output feature maps in each block
                (with 1.0 being the default in the paper)
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:114]: Does this have to be a `Float64`?

```suggestion
function MobileNetv2(width_mult = 1.0; pretrain = false, nclasses = 1000)
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:138]:
```suggestion
    selayer(channels, reduction = 4)

Squeeze and Excitation layer used by MobileNetv3
([reference](https://arxiv.org/abs/1905.02244)).
```

File: src/mobilenet.jl
Comment (darsnack) [reply src/mobilenet.jl:85]: Could make this a vector of tuples for consistency.

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:233]:
```suggestion
  :small => [
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:247]:
```suggestion
  :large => [
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:201]: It doesn't look like `mode` is used anywhere in the function anymore?

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:282]:
```suggestion
- `mode`: :small or :large for the size of the model (see paper).
- `width_mult`: Controls the number of output feature maps in each block
                (with 1.0 being the default in the paper).
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:273]:
```suggestion
    MobileNetv3(mode = :small, width_mult = 1.0; pretrain = false, nclasses = 1000)
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:286]:
```suggestion
function MobileNetv3(mode = :small, width_mult = 1.0; pretrain = false, nclasses = 1000)
```

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl:134]: Thank you so much!

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:287]:
```suggestion
  @assert mode in [:large, :small] "`mode` has to be either :large or :small"
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:289]:
```suggestion
  max_width = (mode == :large) ? 1280 : 1024
```

File: test/runtests.jl
Comment (darsnack) [inline test/runtests.jl:125]:
```suggestion
      m = MobileNetv3(mode)
```

File: test/runtests.jl
Comment (darsnack) [inline test/runtests.jl:131]:
```suggestion
        @test_throws ArgumentError MobileNetv3(mode; pretrain = true)
```

**Review by darsnack — CHANGES_REQUESTED** [review]
Okay, shaping up pretty well. I agree, let's not include the config as part of the arguments for MobileNetV2 for now.

File: src/mobilenet.jl
Comment (darsnack) [reply src/mobilenet.jl]: I don't really think there's a big difference here, but Julia conventions typically use `Symbol`s for arguments of this form. `Symbol`s do have a special meaning in metaprogramming, but in this context, they are effectively just another data type (i.e. we are not doing any metaprogramming).

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl:115]: Nope, realised that the function call works just fine without. Fixed

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl]: Yep, got eliminated amongst the shuffle of the configs

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:115]:
```suggestion
function MobileNetv2(width_mult = 1; pretrain = false, nclasses = 1000)
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:196]: I see the reasoning for `config` first in this case, but I think consistency with v2 outweighs it.
```suggestion
function mobilenetv3(width_mult, configs; max_width = 1024, nclasses = 1000)
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:282]:
```suggestion
function MobileNetv3(mode = :small, width_mult = 1; pretrain = false, nclasses = 1000)
```

**Review by darsnack — CHANGES_REQUESTED** [review]
Few small changes. Otherwise this is good to go.

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl:115]: I kept it as 1.0 so that it's clear to the user that `width_mult` is in general a floating point value. Will change it if you think that's apparent, though

File: src/mobilenet.jl
Comment (theabhirath) [reply src/mobilenet.jl:282]: This one runs into problems when I remove the ::Symbol. It doesn't call the function and just displays as `MobileNetv3(:small)`. Hence the type annotation. Same comment for the `width_mult` as above

File: src/mobilenet.jl
Comment (darsnack) [reply src/mobilenet.jl:115]: Sure though we can do that in the docstring and not the code (that's the part the user will see).

File: src/mobilenet.jl
Comment (darsnack) [reply src/mobilenet.jl:282]: Oh right cause it conflicts with the default constructor for the struct. We can leave them in then, but widen the type of `widen_mult` to `Number` instead of `Float64`.

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:178]:
```suggestion
    mobilenetv3(width_mult, configs; max_width = 1024, nclasses = 1000)
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:192]:
```suggestion

- `width_mult`: Controls the number of output feature maps in each block
                (with 1.0 being the default in the paper;
                 this is usually a value between 0.1 and 1.4)
- `configs`: a "list of tuples" configuration for each layer that details:
  - `k::Int` - The size of the convolutional kernel
  - `c::Float` - The multiplier factor for deciding the number of feature maps in the hidden layer
  - `t::Int` - The number of output feature maps for a given block
  - `use_se::Bool` - Whether to use Squeeze and Excitation layer
  - `use_hs::Bool` - Whether to use Hard-Swish activation function
  - `s::Int` - The stride of the convolutional kernel
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:101]:
```suggestion
    MobileNetv2(width_mult = 1.0; pretrain = false, nclasses = 1000)
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:109]:
```suggestion
- `width_mult`: Controls the number of output feature maps in each block
                (with 1.0 being the default in the paper;
                 this is usually a value between 0.1 and 1.4)
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:51]:
```suggestion

- `width_mult`: Controls the number of output feature maps in each block
                (with 1.0 being the default in the paper;
                 this is usually a value between 0.1 and 1.4)
```

File: src/mobilenet.jl
Comment (darsnack) [inline src/mobilenet.jl:276]:
```suggestion
- `width_mult`: Controls the number of output feature maps in each block
                (with 1.0 being the default in the paper;
                 this is usually a value between 0.1 and 1.4)
```

**Review by darsnack — CHANGES_REQUESTED** [review]

**Review by darsnack — APPROVED** [review]
Looks good, will merge once CI is green! Thanks for another great PR @theabhirath.

**Comment by darsnack** [comment]
@theabhirath some of the old changes were reverted by the force push

**Comment by theabhirath** [comment]
Shoot, must've missed those. The tests were failing because we changed the API for the MobileNetv3 so I had to go back and fix em. This seems fine now though?

**Comment by darsnack** [comment]
Okay just waiting on checks now

**Comment by theabhirath** [comment]
I'm a little confused how this happened, honestly, given that I only squashed my commits, so the earlier changes should've still been preserved

**Comment by darsnack** [comment]
Yeah not sure. Just minor docstring changes anyways.

**Pull request merged**
